mongodbatlas/resource_mongodbatlas_online_archive.go

```diff
@@ -0,0 +1,414 @@
+	if err := d.Set("project_id", projectID); err != nil {
+		log.Printf("error setting project id %s for Online Archive id: %s", err, atlasID)
+	}
+
+	d.SetId(encodeStateID(map[string]string{
+		"atlas_id":     outOnlineArchive.ID,
+		"cluster_name": outOnlineArchive.ClusterName,
+		"project_id":   projectID,
+	}))
+
+	return []*schema.ResourceData{d}, nil
+}
+
+func mapToArchivePayload(d *schema.ResourceData) matlas.OnlineArchive {
+	// shared input
+	requestInput := matlas.OnlineArchive{
+		DBName:   d.Get("db_name").(string),
+		CollName: d.Get("coll_name").(string),
+	}
+
+	requestInput.Criteria = mapCriteria(d)
+
+	if partitions, ok := d.GetOk("partition_fields"); ok {
+		list := partitions.([]interface{})
+
+		if len(list) > 0 {
+			partitionList := make([]*matlas.PartitionFields, 0, len(list))
+			for _, partition := range list {
+				item := partition.(map[string]interface{})
+				localOrder := item["order"].(int)
+				localOrderFloat := float64(localOrder)
+				partitionList = append(partitionList,
+					&matlas.PartitionFields{
+						FieldName: item["field_name"].(string),
+						Order:     pointy.Float64(localOrderFloat),
+					},
+				)
+			}
+
+			requestInput.PartitionFields = partitionList
+		}
+	}
+
+	return requestInput
+}
+
+func resourceMongoDBAtlasOnlineArchiveUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*matlas.Client)
+
+	ids := decodeStateID(d.Id())
+
+	atlasID := ids["atlas_id"]
+	projectID := ids["project_id"]
+	clusterName := ids["cluster_name"]
+
+	// if the criteria or the paused is enable then perform an update
+	paused := d.HasChange("paused")
+	criteria := d.HasChange("criteria")
+
+	// nothing to do, let's go
+	if !paused && !criteria {
+		return nil
+	}
+
+	request := matlas.OnlineArchive{}
+
+	// reading current value
+	if paused {
+		request.Paused = pointy.Bool(d.Get("paused").(bool))
+	}
+
+	if criteria {
+		request.Criteria = mapCriteria(d)
+	}
+
+	_, _, err := conn.OnlineArchives.Update(context.Background(), projectID, clusterName, atlasID, &request)
+
+	if err != nil {
+		return fmt.Errorf("error updating Mongo Online Archive id: %s %s", atlasID, err.Error())
+	}
+
+	return resourceMongoDBAtlasOnlineArchiveRead(d, meta)
+}
+
+func fromOnlineArchiveToMap(in *matlas.OnlineArchive) map[string]interface{} {
+	// computed attribute
+	schemaVals := map[string]interface{}{
+		"cluster_name": in.ClusterName,
+		"atlas_id":     in.ID,
+		"paused":       in.Paused,
+		"state":        in.State,
+		"coll_name":    in.CollName,
+	}
+
+	// criteria
+	criteria := map[string]interface{}{
+		"type":              in.Criteria.Type,
+		"date_field":        in.Criteria.DateField,
+		"date_format":       in.Criteria.DateFormat,
+		"expire_after_days": int(in.Criteria.ExpireAfterDays),
+		// missing query check in client
+	}
+
+	// clean up criteria for empty values
+	for key, val := range criteria {
+		if isEmpty(val) {
+			delete(criteria, key)
+		}
+	}
+
+	schemaVals["criteria"] = criteria
```

Comment:
yea I know it's ugly. yea it's a leftover let me remove code from other parts, the reason because it was in the beginning is that  thesame functions were reused for both datasource and resource. So I had some values sometimes for datasource and others in resource, so I had to manipulate the data. But I updated the client, so no longer needed :)